### PR TITLE
Switch to Encodable for JSONification

### PR DESCRIPTION
**Pull request by TellowKrinkle**
Now that Swift 4 is out, we have Codable!  Yaaaay!
Except that it wasn't meant for situations where you use some structs and some random dictionaries.  It really doesn't like trying to encode `[String: Any]`s or even `[String: Encodable]`s (trying to encode either with the default implementation throws a FatalError saying `fatal error: Array<Encodable> does not conform to Encodable because Encodable does not conform to itself. You must use a concrete type to encode or decode.`)
Is the new implementation hackish?  A little.  But I think it's a little less hackish than the one it's replacing.  Also, we can finally follow the Swift API Guidelines on acronyms (I'm talking about you, `Id` and `Url`) without an overly complicated snakecase function (though that's not implemented here).

### CODE REVIEW

**Comment by nuclearace** [comment]
Yeah whenever conditional conformance arrives it should get rid of the hacky things we have to here since it'll let them handle these in the stdlib

**Comment by nuclearace** [comment]
@tellowkrinkle Have you toyed around with trying bring deserialization to the holding types? It would be fantastic if we could get rid of all that crufty stuff we have now.

File: Sources/SwiftDiscord/DiscordSnowflakeID.swift [hunk 1]
```diff
@@ -108,12 +108,32 @@ extension Snowflake {
...
+    /// Set this key to true on an encoder to encode Snowflakes as UInt64s instead of Strings.
+    /// This will save space in binary encoders like binary plists, but will cause encoded JSON to be
+    /// incompatible with Discord. Since JSON isn't a binary encoding, it won't save much space there anyways.
+    public static let encodeAsUInt64 = CodingUserInfoKey(rawValue: "snowflakeAsUInt64")!
+    public init(from decoder: Decoder) throws {
+        do {
+            let intForm = try UInt64(from: decoder)
+            self = Snowflake(intForm)
+        }
+        catch _ {
```
Comment (nuclearace) [inline Sources/SwiftDiscord/DiscordSnowflakeID.swift]: Catch on same line as }. And you should probably log the error. The _ is redundant right? You can leave it out and you'll get an implicit `error` variable.

File: Sources/SwiftDiscord/DiscordJSON.swift [hunk 2]
```diff
@@ -21,15 +21,24 @@ enum JSON {
...
-    static func encodeJSONData(_ object: Any) -> Data? {
-        return try? JSONSerialization.data(withJSONObject: object)
+    static func encodeJSONData<T: Encodable>(_ object: T) -> Data? {
+        let encoder = JSONEncoder()
+        encoder.dateEncodingStrategy = .formatted(DiscordDateFormatter.rfc3339DateFormatter)
+        if let dictionary = object as? [String: Any] {
+            guard let encodableDic = dictionary as? [String: Encodable] else { return nil }
+            return try? encoder.encode(GenericEncodableDictionary(encodableDic))
+        }
+        if let array = object as? [Any] {
```
Comment (nuclearace) [inline Sources/SwiftDiscord/DiscordJSON.swift]: `else if let` reads better to me.

File: Sources/SwiftDiscord/DiscordJSON.swift [hunk 3]
```diff
@@ -74,117 +83,3 @@ enum JSONError : Error {
...
-    guard valueMirror.children.count == 1 else { return NSNull() } // If no children, this is .none
-    // Make sure the wrapped type is representable in json
-    guard case let (_, optionalValue as JSONRepresentable) = valueMirror.children.first! else { return nil }
-
-    if let jsonable = optionalValue as? JSONAble {
-        return jsonable.json
-    }
-
-    return try? optionalValue.jsonValue()
-}
```
Comment (nuclearace) [inline Sources/SwiftDiscord/DiscordJSON.swift:190]: 👍

File: Sources/SwiftDiscord/DiscordSnowflakeID.swift [hunk 4]
```diff
@@ -108,12 +108,32 @@ extension Snowflake {
...
+            let stringForm = try String(from: decoder)
```
Comment (nuclearace) [inline Sources/SwiftDiscord/DiscordSnowflakeID.swift]: If you only use a variable once, you can toss it and move it where it's used. Some exceptions are for expressions that getting busy or when you're at a line length limit (I use 120).

File: Sources/SwiftDiscord/DiscordSnowflakeID.swift [hunk 5]
```diff
@@ -108,12 +108,32 @@ extension Snowflake {
...
+            guard let snowflake = Snowflake(stringForm) else {
+                throw DecodingError.typeMismatch(Snowflake.self, DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Failed to convert decoded string into a snowflake"))
```
Comment (nuclearace) [inline Sources/SwiftDiscord/DiscordSnowflakeID.swift]: Line length.

**Review by nuclearace — CHANGES_REQUESTED** [review]
This is good, something I wanted to do anyway. I'm fine with the bit hacky stuff we have to do with the container types. I anticipate that when Swift 5 drops with conditional conformances (knock on wood) this can go away since it'll let the stdlib do `extension Array : Codeable where Element : Codeable`

**Comment by nuclearace** [comment]
Although thinking about it more I guess conditional conformance really won't help with the case where it's just Any... since it can't know what to encode. And toying around it doesn't look like you can define a custom protocol that just declares Codable conformance to use, which is kind of a bummer.

**Comment by TellowKrinkle** [comment]
I haven't looked into deserialization much yet.  I do know that Codable doesn't support autogenerating with defaults, so we'll still need to manually implement `init(from decoder:)`.

**Pull request merged**